source/Body.cpp

```diff
@@ -352,10 +352,10 @@ void Body::Turn(double amount)
 		return;
 
 	auto RotatePointAroundOrigin = [](Point &toRotate, double radians) -> Point {
-		float si = sin(radians);
-		float co = cos(radians);
-		float newX = toRotate.X() * co - toRotate.Y() * si;
-		float newY = toRotate.X() * si + toRotate.Y() * co;
+		double si = sin(radians);
+		double co = cos(radians);
+		double newX = toRotate.X() * co - toRotate.Y() * si;
+		double newY = toRotate.X() * si + toRotate.Y() * co;
```

Comment:
just to make this clear, this is no optimization, quite the opposite

Reply:
I disagree.

`sin()` returns a double, which is converted to a float to assign to `si`. same with `cos()`. these floats are then converted back to doubles to perform the multiplications on the following two lines (`.X()` and `.Y()` return doubles, and a binary operand with a float and a double is upgraded to a double and a double). The resultant doubles are then added/subtracted, resulting in another double, which is then converted to a float to assign to `newX` and `newY`.

Therefore, all the mathematical operations happen in double-precision land, but the stack storage is going through a float conversion and back. There are four conversions from double-precision to single-precision, and four from single back to double.

By changing the declaration, all the unnecessary conversions to/from single-precision are all omitted. It does not change any of the mathematical operations (except perhaps some precision is retained).

Reply:
> these floats are then converted back to doubles to perform the multiplications on the following two lines (.X() and .Y() return doubles, and a binary operand with a float and a double is upgraded to a double and a double).

Actually, that's not entirely true. I did some micro-benchmarking, and the results were the exact same. The compiler is smart enough to realize that those floats only get used in a double-precision context, so it seems to omit the casts. (Also, float->double is extremely efficient in hardware, since it can be done by copies.)

According to the [specification](https://en.cppreference.com/w/cpp/language/operator_arithmetic):
> Floating-point environment
>
> If[ #pragma STDC FENV_ACCESS](https://en.cppreference.com/w/cpp/preprocessor/impl) is supported and set to ON, all floating-point arithmetic operators obey the current floating-point [rounding direction](https://en.cppreference.com/w/cpp/numeric/fenv/FE_round) and report floating-point arithmetic errors as specified in [math_errhandling](https://en.cppreference.com/w/cpp/numeric/math/math_errhandling) unless part of a [static initializer](https://en.cppreference.com/w/cpp/language/initialization#Non-local_variables) (in which case floating-point exceptions are not raised and the rounding mode is to nearest).
>
>Floating-point contraction
>
> Unless[ #pragma STDC FP_CONTRACT](https://en.cppreference.com/w/cpp/preprocessor/impl) is supported and set to OFF, all floating-point arithmetic may be performed as if the intermediate results have infinite range and precision, that is, optimizations that omit rounding errors and floating-point exceptions are allowed. For example, C++ allows the implementation of (x * y) + z with a single fused multiply-add CPU instruction or optimization of a = x * x * x * x; as tmp = x * x; a = tmp * tmp.
> 
> Unrelated to contracting, intermediate results of floating-point arithmetic may have range and precision that is different from the one indicated by its type, see [FLT_EVAL_METHOD](https://en.cppreference.com/w/cpp/types/climits/FLT_EVAL_METHOD).
>
> Formally, the C++ standard makes no guarantee on the accuracy of floating-point operations.

Mind you, the specs do say that values get converted:
> Otherwise, if the [floating-point conversion ranks](https://en.cppreference.com/w/cpp/language/usual_arithmetic_conversions#Floating-point_conversion_rank) of the types of the operands are ordered but(since C++23) not equal, then the operand of the type with the lesser floating-point conversion rank is converted to the type of the other operand.

Except the conversion can freely be optimized out, and there is nothing stopping the compiler from defining asymmetric overloads for arithmetic operators. (If I read it correctly, the specification doesn't actually say that the built-in operators are symmetric, just that their values get converted to the same fp rank, and even that's only for the guaranteed built-in operators.)

Btw, since there is no guarantee made about the accuracy of the operation, the compiler could technically cheat and do the whole operation with floats. Not that it would matter for performance.